src/chainrules.jl

```diff
@@ -0,0 +1,69 @@
+@non_differentiable validate_woodbury_arguments(A, D, S)
+
+# Rule for Woodbury * Real. 
+# Ignoring Complex version for now. 
+function ChainRulesCore.rrule(::typeof(*), A::WoodburyPDMat, B::Real)
```

Comment:
Sorry to hop in late here, but why do we need a rule for this, as opposed to just opting out?

In particular, our implementation of this function ought to be straightforwardly differentiable by Zygote et al:
```julia
*(a::WoodburyPDMat, c::Real) = WoodburyPDMat(a.A, a.D * c, a.S * c)
*(c::Real, a::WoodburyPDMat) = a * c
```

Is it not sufficient just to `@opt_out`?

Reply:
Yeah this is probabaly true. AFAIK this works - here however I was using this to test where the primal returns a woodbury (and thus your tangent is woodbury) as I wanted to use this as a basis to sort out how to work on the rule for `*(H::Diagonal, S::WoodburyPDMat)`. If you remember we had issues with that and had to do `WoodburyPDMat(H * S.A, S.D, H * S.S * H' + D)`. This was actually the direction I wanted to take this MR, but  possibly a lot has moved on since that issue. 

I hadn't considered opt-outs... if the rule works are they generally not more efficient than zygote working it out?

Reply:
> Yeah this is probabaly true. AFAIK this works - here however I was using this to test where the primal returns a woodbury (and thus your output co-tangent is a tangent{woodbury}) as I wanted to use this as a basis to sort out how to work on the rule for *(H::Diagonal, S::WoodburyPDMat). If you remember we had issues with that and had to do WoodburyPDMat(H * S.A, S.D, H * S.S * H' + D). This was actually the direction I wanted to take this MR, but possibly a lot has moved on since that issue.

Interesting -- again, I'd suggest just opting out here 🤷 Or am I missing the point?

> I hadn't considered opt-outs... if the rule works are they generally not more efficient than zygote working it out?

No. In simple cases like this there's no reason that Zygote can't achieve optimal performance. If you have control flow (either via a for-loop or an `if-else-end` block) Zygote will generally struggle, but otherwise you're definitely better off letting Zygote figure stuff out.

My experience has consistently been that if you have the choice between writing a rule and writing AD-friendly code, always pick the latter.